### PR TITLE
fix: EntitlementIterator behavior and type-hinting

### PR DESCRIPTION
**Pull request by yoggys**
## Summary

EntitlementIterator now behaves correctly when using after, before and limit (`=None` or `>100`) parameters.
https://canary.discord.com/channels/881207955029110855/1274835760842801184

What exactly does it fix?
- when using the after parameter the data is sorted by ID in ascending order
- when using the before parameter, or not using any, the data are sorted by ID decreasingly

The previous code set: `self.after = Object(id=int(data[-1][“id”]))`.
Which only allowed to use after, using before or nothing resulted in using the smaller ID as after so creating duplicate records (only when requesing for more pages so limit > 100).

## Information

- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [ ] I have updated the changelog to include these changes.

### CODE REVIEW

File: discord/iterators.py [hunk 1]
```diff
@@ -1014,30 +1025,57 @@ async def fill_entitlements(self):
...
             before=before,
-            after=after,
-            limit=self.retrieve,
+            limit=retrieve,
             user_id=self.user_id,
             guild_id=self.guild_id,
             sku_ids=self.sku_ids,
             exclude_ended=self.exclude_ended,
         )
+        if len(data):
```
Comment (plun1331) [inline discord/iterators.py:1057]: This could be simplified to `if data`

File: discord/iterators.py [hunk 2]
```diff
@@ -1014,30 +1025,57 @@ async def fill_entitlements(self):
...
+        data = await self.get_entitlements(
+            self.state.application_id,
+            after=after,
+            limit=retrieve,
+            user_id=self.user_id,
+            guild_id=self.guild_id,
+            sku_ids=self.sku_ids,
+            exclude_ended=self.exclude_ended,
+        )
+        if len(data):
```
Comment (plun1331) [inline discord/iterators.py:1077]: Same as above

**Comment by yoggys** [comment]
> I believe the entitlements may need sorting based on what method is used to request, they might be backwards for each request if requested with `after` (or maybe `before`, the other iterators should have some implementation of reversing the thing they're fetching)

yeah, right now it just fixes issue for fetching data but data order depends on method (ASC/DESC)

**Comment by yoggys** [comment]
Okay so I just tested it again and for debug i set limit as 5 to check behavior of iterator:

- Before + After = OK DESC
image

- Before = OK DESC
image

- After = OK ASC
image

**Review by plun1331 — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**